File: skills/pretext/scripts/run_pretext_review_iteration.py
#!/usr/bin/env python3
"""Run one review iteration for the Pretext skill using Claude CLI."""

from __future__ import annotations

import argparse
import json
import os
import re
import subprocess
import time
from dataclasses import dataclass
from pathlib import Path


DEFAULT_EVAL_IDS = [1, 4, 5, 8, 9, 11, 24]


@dataclass(frozen=True)
class EvalItem:
    eval_id: int
    prompt: str
    expected_output: str
    expectations: list[str]


def slugify(text: str, limit: int = 48) -> str:
    lowered = text.lower()
    slug = re.sub(r"[^a-z0-9]+", "-", lowered).strip("-")
    if not slug:
        return "eval"
    return slug[:limit].rstrip("-")


def load_evals(path: Path) -> dict[int, EvalItem]:
    data = json.loads(path.read_text(encoding="utf-8"))
    mapping: dict[int, EvalItem] = {}
    for item in data["evals"]:
        mapping[item["id"]] = EvalItem(
            eval_id=item["id"],
            prompt=item["prompt"],
            expected_output=item["expected_output"],
            expectations=item["expectations"],
        )
    return mapping


def build_prompt(eval_item: EvalItem, skill_path: Path | None) -> str:
    shared = (
        "Answer the task in concise technical Markdown.\n"
        "State the chosen route explicitly when relevant.\n"
        "Do not mention this evaluation harness.\n\n"
        f"Task:\n{eval_item.prompt}\n"
    )
    if skill_path is None:
        return (
            "You are answering a technical task directly without any external skill instructions.\n\n"
            + shared
        )
    return (
        "You are evaluating a local Claude-style skill.\n"
        f"Skill path: {skill_path}\n\n"
        "Required process:\n"
        f"1. Read {skill_path / 'SKILL.md'} first.\n"
        "2. Follow the skill faithfully.\n"
        "3. Load only the minimum referenced files needed for the task.\n"
        "4. Answer the task directly.\n\n"
        + shared
    )


def run_claude(prompt: str, workdir: Path, model: str | None) -> tuple[str, str, int, float]:
    cmd = [
        "claude",
        "-p",
        prompt,
        "--output-format",
        "text",
        "--permission-mode",
        "bypassPermissions",
    ]
    if model:
        cmd.extend(["--model", model])

    env = {k: v for k, v in os.environ.items() if k != "CLAUDECODE"}
    start = time.time()
    result = subprocess.run(
        cmd,
        cwd=workdir,
        capture_output=True,
        text=True,
        encoding="utf-8",
        errors="replace",
    )
    elapsed = time.time() - start
    return result.stdout, result.stderr, result.returncode, elapsed


def write_json(path: Path, payload: object) -> None:
    path.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")


def run_iteration(
    eval_set_path: Path,
    skill_path: Path,
    workspace: Path,
    eval_ids: list[int],
    model: str | None,
) -> None:
    evals = load_evals(eval_set_path)
    workspace.mkdir(parents=True, exist_ok=True)

    history = {
        "iteration": workspace.name,
        "skill_path": str(skill_path),
        "eval_ids": eval_ids,
        "started_at_epoch": time.time(),
    }
    write_json(workspace / "iteration_metadata.json", history)

    for eval_id in eval_ids:
        if eval_id not in evals:
            raise KeyError(f"Unknown eval id {eval_id}")
        item = evals[eval_id]
        eval_slug = slugify(item.expected_output)
        eval_dir = workspace / f"eval-{eval_id:02d}-{eval_slug}"
        eval_dir.mkdir(parents=True, exist_ok=True)
        write_json(
            eval_dir / "eval_metadata.json",
            {
                "eval_id": item.eval_id,
                "eval_name": f"eval-{item.eval_id:02d}-{eval_slug}",
                "prompt": item.prompt,
                "expectations": item.expectations,
                "expected_output": item.expected_output,
            },
        )

        configs = [
            ("with_skill", skill_path),
            ("without_skill", None),
        ]
        repo_root = skill_path.parents[1]
        for config_name, config_skill_path in configs:
            run_dir = eval_dir / config_name / "run-1"
            outputs_dir = run_dir / "outputs"
            outputs_dir.mkdir(parents=True, exist_ok=True)

            prompt = build_prompt(item, config_skill_path)
            stdout, stderr, returncode, elapsed = run_claude(prompt, workdir=repo_root, model=model)
            answer = stdout.strip()
            if returncode != 0 and not answer:
                answer = f"Execution failed with return code {returncode}.\n\n{stderr.strip()}"

            transcript = (
                f"# Eval Transcript\n\n"
                f"## Eval Prompt\n\n{item.prompt}\n\n"
                f"## Harness Prompt\n\n```\n{prompt}\n```\n\n"
                f"## Result\n\n{returncode}\n\n"
                f"## Stdout\n\n{answer}\n\n"
                f"## Stderr\n\n{stderr.strip()}\n"
            )

            (outputs_dir / "answer.md").write_text(answer + ("\n" if not answer.endswith("\n") else ""), encoding="utf-8")
            (run_dir / "transcript.md").write_text(transcript, encoding="utf-8")
            (outputs_dir / "user_notes.md").write_text(
                (
                    f"- configuration: `{config_name}`\n"
                    f"- skill_path: `{config_skill_path}`\n"
                    f"- return_code: `{returncode}`\n"
                ),
                encoding="utf-8",
            )
            write_json(
                outputs_dir / "metrics.json",
                {
                    "tool_calls": {},
                    "total_tool_calls": 0,
                    "total_steps": 0,
                    "files_created": ["answer.md"],
                    "errors_encountered": 0 if returncode == 0 else 1,
                    "output_chars": len(answer),
                    "transcript_chars": len(transcript),
                },
            )
            write_json(
                run_dir / "timing.json",
                {
                    "total_tokens": 0,
                    "duration_ms": int(elapsed * 1000),
                    "total_duration_seconds": round(elapsed, 3),
                    "executor_duration_seconds": round(elapsed, 3),
                },
            )


def main() -> int:
    parser = argparse.ArgumentParser(description="Run one Pretext review iteration with Claude CLI.")
    parser.add_argument(
        "--eval-set",
        default="skills/pretext/evals/evals.json",
        help="Path to evals.json.",
    )
    parser.add_argument(
        "--skill-path",
        default="skills/pretext",
        help="Path to the Pretext skill directory.",
    )
    parser.add_argument(
        "--workspace",
        default="skills/pretext-workspace/iteration-1",
        help="Workspace directory for the iteration.",
    )
    parser.add_argument(
        "--eval-id",
        action="append",
        type=int,
        help="Eval ID to run. Repeatable. Defaults to a representative first-iteration subset.",
    )
    parser.add_argument("--model", default=None, help="Optional Claude model override.")
    args = parser.parse_args()

    eval_ids = args.eval_id or DEFAULT_EVAL_IDS
    run_iteration(
        eval_set_path=Path(args.eval_set).resolve(),
        skill_path=Path(args.skill_path).resolve(),
        workspace=Path(args.workspace).resolve(),
        eval_ids=eval_ids,
        model=args.model,
    )
    print(f"Review iteration written to {Path(args.workspace).resolve()}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/pretext/scripts/run-pretext-review-iteration.ts
#!/usr/bin/env bun
/**
 * Run one review iteration for the Pretext skill using Claude CLI.
 *
 * Usage:
 *   bun run skills/pretext/scripts/run-pretext-review-iteration.ts
 *   bun run skills/pretext/scripts/run-pretext-review-iteration.ts --eval-id 1 --eval-id 5
 *   bun run skills/pretext/scripts/run-pretext-review-iteration.ts --model claude-sonnet-4-20250514
 */

import { resolve, join } from "node:path";
import { parseArgs } from "node:util";
import { mkdirSync, existsSync } from "node:fs";

// ── Constants ──────────────────────────────────────────────────────────────

const DEFAULT_EVAL_IDS = [1, 4, 5, 8, 9, 11, 24];

// ── Types ──────────────────────────────────────────────────────────────────

interface EvalItem {
  evalId: number;
  prompt: string;
  expectedOutput: string;
  expectations: string[];
}

interface EvalsData {
  evals: {
    id: number;
    prompt: string;
    expected_output: string;
    expectations: string[];
  }[];
}

// ── Helpers ────────────────────────────────────────────────────────────────

function slugify(text: string, limit = 48): string {
  const slug = text
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-|-$/g, "");
  if (!slug) return "eval";
  return slug.slice(0, limit).replace(/-$/, "");
}

function loadEvals(data: EvalsData): Map<number, EvalItem> {
  const mapping = new Map<number, EvalItem>();
  for (const item of data.evals) {
    mapping.set(item.id, {
      evalId: item.id,
      prompt: item.prompt,
      expectedOutput: item.expected_output,
      expectations: item.expectations,
    });
  }
  return mapping;
}

function buildPrompt(evalItem: EvalItem, skillPath: string | null): string {
  const shared =
    "Answer the task in concise technical Markdown.\n" +
    "State the chosen route explicitly when relevant.\n" +
    "Do not mention this evaluation harness.\n\n" +
    `Task:\n${evalItem.prompt}\n`;

  if (skillPath === null) {
    return (
      "You are answering a technical task directly without any external skill instructions.\n\n" +
      shared
    );
  }

  const skillMdPath = join(skillPath, "SKILL.md");
  return (
    "You are evaluating a local Claude-style skill.\n" +
    `Skill path: ${skillPath}\n\n` +
    "Required process:\n" +
    `1. Read ${skillMdPath} first.\n` +
    "2. Follow the skill faithfully.\n" +
    "3. Load only the minimum referenced files needed for the task.\n" +
    "4. Answer the task directly.\n\n" +
    shared
  );
}

async function runClaude(
  prompt: string,
  workdir: string,
  model: string | undefined,
): Promise<{ stdout: string; stderr: string; exitCode: number; elapsed: number }> {
  const cmd = [
    "claude",
    "-p",
    prompt,
    "--output-format",
    "text",
    "--permission-mode",
    "bypassPermissions",
  ];
  if (model) {
    cmd.push("--model", model);
  }

  // Filter out CLAUDECODE env var
  const env: Record<string, string> = {};
  for (const [k, v] of Object.entries(process.env)) {
    if (k !== "CLAUDECODE" && v !== undefined) {
      env[k] = v;
    }
  }

  const start = performance.now();
  const proc = Bun.spawn(cmd, {
    cwd: workdir,
    env,
    stdout: "pipe",
    stderr: "pipe",
  });

  const [stdout, stderr] = await Promise.all([
    new Response(proc.stdout).text(),
    new Response(proc.stderr).text(),
  ]);
  const exitCode = await proc.exited;
  const elapsed = (performance.now() - start) / 1000;

  return { stdout, stderr, exitCode, elapsed };
}

async function writeJson(path: string, payload: unknown): Promise<void> {
  await Bun.write(path, JSON.stringify(payload, null, 2) + "\n");
}

// ── Core iteration ─────────────────────────────────────────────────────────

async function runIteration(
  evalSetPath: string,
  skillPath: string,
  workspace: string,
  evalIds: number[],
  model: string | undefined,
): Promise<void> {
  const evalsData: EvalsData = await Bun.file(evalSetPath).json();
  const evals = loadEvals(evalsData);

  mkdirSync(workspace, { recursive: true });

  const history = {
    iteration: workspace.split(/[\\/]/).pop(),
    skill_path: skillPath,
    eval_ids: evalIds,
    started_at_epoch: Date.now() / 1000,
  };
  await writeJson(join(workspace, "iteration_metadata.json"), history);

  // Resolve repo root from skill path (skill_path is …/skills/pretext → 2 parents up)
  const repoRoot = resolve(skillPath, "..", "..");

  for (const evalId of evalIds) {
    const item = evals.get(evalId);
    if (!item) {
      throw new Error(`Unknown eval id ${evalId}`);
    }

    const evalSlug = slugify(item.expectedOutput);
    const evalDir = join(workspace, `eval-${String(evalId).padStart(2, "0")}-${evalSlug}`);
    mkdirSync(evalDir, { recursive: true });

    await writeJson(join(evalDir, "eval_metadata.json"), {
      eval_id: item.evalId,
      eval_name: `eval-${String(item.evalId).padStart(2, "0")}-${evalSlug}`,
      prompt: item.prompt,
      expectations: item.expectations,
      expected_output: item.expectedOutput,
    });

    const configs: [string, string | null][] = [
      ["with_skill", skillPath],
      ["without_skill", null],
    ];

    for (const [configName, configSkillPath] of configs) {
      const runDir = join(evalDir, configName, "run-1");
      const outputsDir = join(runDir, "outputs");
      mkdirSync(outputsDir, { recursive: true });

      const prompt = buildPrompt(item, configSkillPath);
      const { stdout, stderr, exitCode, elapsed } = await runClaude(
        prompt,
        repoRoot,
        model,
      );
      let answer = stdout.trim();
      if (exitCode !== 0 && !answer) {
        answer = `Execution failed with return code ${exitCode}.\n\n${stderr.trim()}`;
      }

      const transcript =
        `# Eval Transcript\n\n` +
        `## Eval Prompt\n\n${item.prompt}\n\n` +
        `## Harness Prompt\n\n\`\`\`\n${prompt}\n\`\`\`\n\n` +
        `## Result\n\n${exitCode}\n\n` +
        `## Stdout\n\n${answer}\n\n` +
        `## Stderr\n\n${stderr.trim()}\n`;

      await Bun.write(
        join(outputsDir, "answer.md"),
        answer + (answer.endsWith("\n") ? "" : "\n"),
      );
      await Bun.write(join(runDir, "transcript.md"), transcript);
      await Bun.write(
        join(outputsDir, "user_notes.md"),
        `- configuration: \`${configName}\`\n` +
          `- skill_path: \`${configSkillPath}\`\n` +
          `- return_code: \`${exitCode}\`\n`,
      );
      await writeJson(join(outputsDir, "metrics.json"), {
        tool_calls: {},
        total_tool_calls: 0,
        total_steps: 0,
        files_created: ["answer.md"],
        errors_encountered: exitCode === 0 ? 0 : 1,
        output_chars: answer.length,
        transcript_chars: transcript.length,
      });
      await writeJson(join(runDir, "timing.json"), {
        total_tokens: 0,
        duration_ms: Math.round(elapsed * 1000),
        total_duration_seconds: Math.round(elapsed * 1000) / 1000,
        executor_duration_seconds: Math.round(elapsed * 1000) / 1000,
      });
    }
  }
}

// ── CLI ────────────────────────────────────────────────────────────────────

async function main(): Promise<number> {
  const { values } = parseArgs({
    options: {
      "eval-set": { type: "string", default: "skills/pretext/evals/evals.json" },
      "skill-path": { type: "string", default: "skills/pretext" },
      workspace: { type: "string", default: "skills/pretext-workspace/iteration-1" },
      "eval-id": { type: "string", multiple: true },
      model: { type: "string" },
    },
    strict: true,
  });

  const evalIds =
    values["eval-id"] && values["eval-id"].length > 0
      ? values["eval-id"].map((id) => parseInt(id, 10))
      : DEFAULT_EVAL_IDS;

  const evalSetPath = resolve(values["eval-set"]!);
  const skillPath = resolve(values["skill-path"]!);
  const workspace = resolve(values.workspace!);

  await runIteration(evalSetPath, skillPath, workspace, evalIds, values.model);
  console.log(`Review iteration written to ${workspace}`);
  return 0;
}

process.exit(await main());
